### PR TITLE
Track the evaluation loop outputs in the loop

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

Follow-up to #10810 and #10923

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [n/a] Did you make sure to **update the documentation** with your changes? (if necessary)
- [n/a] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [n/a] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

### CODE REVIEW

File: pytorch_lightning/loops/dataloader/evaluation_loop.py [hunk 1]
```diff
@@ -32,6 +34,7 @@ def __init__(self) -> None:
 
         self._results = ResultCollection(training=False)
         self._outputs: List[EPOCH_OUTPUT] = []
+        self._logged_outputs: List[_OUT_DICT] = []
```
Comment (carmocca) [inline pytorch_lightning/loops/dataloader/evaluation_loop.py:37]: This substitutes the `eval_loop_results` that were in the `LoggerConnector`

File: pytorch_lightning/loops/dataloader/evaluation_loop.py [hunk 2]
```diff
@@ -118,26 +122,49 @@ def advance(self, *args: Any, **kwargs: Any) -> None:
             # indicate the loop has run
             self._has_run = True
 
+    def on_advance_end(self) -> None:
+        # inform logger the epoch loop has finished
+        self.trainer.logger_connector.epoch_end_reached()
+
+        self._logged_outputs.append(self.trainer.logger_connector.update_eval_epoch_metrics())
```
Comment (carmocca) [inline pytorch_lightning/loops/dataloader/evaluation_loop.py:129]: Now, these are computed when each dataloader ends, instead of when all dataloaders end

File: pytorch_lightning/loops/dataloader/evaluation_loop.py [hunk 3]
```diff
@@ -118,26 +122,49 @@ def advance(self, *args: Any, **kwargs: Any) -> None:
...
         self._evaluation_epoch_end(outputs)
 
         # hook
         self._on_evaluation_epoch_end()
 
-        # log epoch metrics
-        eval_loop_results = self.trainer.logger_connector.update_eval_epoch_metrics()
+        logged_outputs, self._logged_outputs = self._logged_outputs, []  # free memory
+        # include any logged outputs on epoch_end
+        if self.num_dataloaders < 2:  # TODO: remove this check
```
Comment (carmocca) [inline pytorch_lightning/loops/dataloader/evaluation_loop.py:144]: Removing this `if` should fix #10904. Left for a separate PR

File: pytorch_lightning/loops/dataloader/evaluation_loop.py [hunk 4]
```diff
@@ -237,3 +261,18 @@ def _on_evaluation_epoch_end(self) -> None:
         self.trainer.call_hook(hook_name)
         self.trainer.call_hook("on_epoch_end")
         self.trainer.logger_connector.on_epoch_end()
+
+    def _print_results(self, results: List[_OUT_DICT], stage: RunningStage) -> None:
+        # TODO: this could be updated to look nicer
```
Comment (carmocca) [inline pytorch_lightning/loops/dataloader/evaluation_loop.py:266]: This was inside `LoggerConnector` before. I'd say it fits much better here, as it's a responsibility of the loop.

I've also added this TODO because we could make it look nicer, or use `rich` if possible.

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py [hunk 5]
```diff
@@ -150,58 +148,19 @@ def update_eval_step_metrics(self) -> None:
...
-
-        on_step = not self._epoch_end_reached
-        num_dataloaders = self.trainer._evaluation_loop.num_dataloaders
-        has_been_initialized = len(self.eval_loop_results) == num_dataloaders
-        assert self.trainer._evaluation_loop._results is not None
-        for dl_idx in range(num_dataloaders):
-            metrics = self.trainer._evaluation_loop._results.metrics(
-                on_step, dataloader_idx=dl_idx if num_dataloaders > 1 else None
-            )
-            callback_metrics = metrics["callback"]
```
Comment (carmocca) [inline pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:165]: This no longer needs to recompute the metrics per dataloader even though we had already computed them in `update_eval_epoch_metrics`

File: pytorch_lightning/loops/dataloader/evaluation_loop.py [hunk 6]
```diff
@@ -118,26 +122,49 @@ def advance(self, *args: Any, **kwargs: Any) -> None:
...
         # enable train mode again
         self._on_evaluation_model_train()
 
-        return eval_loop_results
+        if (
+            self.trainer.state.fn not in (TrainerFn.FITTING, TrainerFn.TUNING)
+            and not self.trainer.sanity_checking
+            and self.trainer.is_global_zero
+            # TODO: this should be defined in this loop, not the Trainer
+            and self.trainer.verbose_evaluate
```
Comment (carmocca) [inline pytorch_lightning/loops/dataloader/evaluation_loop.py:163]: Now that this is here, we can move it from the Trainer to this loop

Comment (carmocca) [reply pytorch_lightning/loops/dataloader/evaluation_loop.py:165]: Waiting in https://github.com/PyTorchLightning/pytorch-lightning/pull/10931

File: pytorch_lightning/loops/dataloader/evaluation_loop.py (hunk 2 above)
Comment (rohitgr7) [reply pytorch_lightning/loops/dataloader/evaluation_loop.py:127]: quick ques: where are taking care of the condition that this method will only compute metrics for the current dataloader_idx?

Comment (carmocca) [reply pytorch_lightning/loops/dataloader/evaluation_loop.py:127]: This will call `ResultCollection.metrics(False)` which in turn calls `ResultCollection.valid_items()` where we filter with the `ResultCollection.dataloader_idx` value

The LoggerConnector makes sure the `ResultCollection.dataloader_idx` value is the correct one

**Review by rohitgr7 — APPROVED** [review]
this is great!

File: pytorch_lightning/loops/dataloader/evaluation_loop.py (hunk 2 above)
Comment (carmocca) [reply pytorch_lightning/loops/dataloader/evaluation_loop.py:127]: > why is there so much space between lines? minimal distance rule for covid does not apply for code.

It's the general style used int he rest of this file. I also prefer tighter code. Feel free to push whitespace changes to your liking!

**Review by awaelchli — APPROVED** [review]

**Comment by carmocca** [comment]
@tchaton this needs your review :)

File: pytorch_lightning/loops/dataloader/evaluation_loop.py (hunk 2 above)
Comment (tchaton) [reply pytorch_lightning/loops/dataloader/evaluation_loop.py:127]: Also, I think this might create some metric problems when doing cross-dataloaders aggregation as the first compute would be done on the first datalaoder and _compute would be stored.

Here is the failing test:

```py
def test_multiple_dataloaders_logging(tmpdir):

    class TestModel(BoringModel):

        def validation_step(self, batch, batch_idx, dataloader_idx):
            self.log("value_1", dataloader_idx, add_dataloader_idx=False)
            self.log("value_2", dataloader_idx, add_dataloader_idx=True)

        def val_dataloader(self):
            return [self.train_dataloader(), self.train_dataloader()]

    model = TestModel()
    model.validation_epoch_end = None
    trainer = Trainer(default_root_dir=tmpdir)
    results = trainer.validate(model)
    assert results == [{'value_2/dataloader_idx_0': 0.0, 'value_1': 0.5}, {'value_2/dataloader_idx_1': 1.0, 'value_1': 0.5}]
```

**Review by tchaton — CHANGES_REQUESTED** [review]
Hey @carmocca, some investigation required around `add_dataloader_idx=False`. I added a test.

File: pytorch_lightning/loops/dataloader/evaluation_loop.py (hunk 2 above)
Comment (tchaton) [reply pytorch_lightning/loops/dataloader/evaluation_loop.py:127]: Here is a PR to discuss the right behaviour: https://github.com/PyTorchLightning/pytorch-lightning/pull/11125/files

Comment (tchaton) [reply pytorch_lightning/loops/dataloader/evaluation_loop.py:127]: This would need better investigation, but it seems this is confusing. I will open an issue and unblock this PR.

**Review by tchaton — APPROVED** [review]
